audio-worklet/design-pattern/lib/wa2.js

```diff
@@ -0,0 +1,289 @@
+    console.assert(!this._isInitialized);
+
+    const channelByteSize = this._length * WA2.BYTES_PER_SAMPLE;
+    const dataByteSize = this._channelCount * channelByteSize;
+    this._dataPtr = Module._malloc(dataByteSize);
+    this._channelData = [];
+    for (let i = 0; i < this._channelCount; ++i) {
+      let startByteOffset = this._dataPtr + i * channelByteSize
+      let endByteOffset = startByteOffset + channelByteSize;
+      // Get the actual array index by dividing the byte offset by 2 bytes.
+      this._channelData[i] =
+          Module.HEAPF32.subarray(startByteOffset >> WA2.BYTES_PER_UNIT,
+                                  endByteOffset >> WA2.BYTES_PER_UNIT);
+    }
+  }
+
+  /**
+   * Adapt the current channel count to the new input buffer.
+   *
+   * @param  {number} newChannelCount The new channel count.
+   */
+  adaptChannel(newChannelCount) {
+    console.assert(this._isInitialized);
+    console.assert(newChannelCount < this._maxChannelCount);
+
+    if (this._channelCount !== newChannelCount) {
+      this._channelCount = newChannelCount;
+    }
+  }
+
+  /**
+   * Getter for the buffer length in frames.
+   *
+   * @return {number} Buffer length in frames.
+   */
+  get length() {
+    return this._isInitialized ? this._length : null;
+  }
+
+  /**
+   * Getter for the number of channels.
+   *
+   * @return {number} Buffer length in frames.
+   */
+  get numberOfChannels() {
+    return this._isInitialized ? this._channelCount : null;
+  }
+
+  /**
+   * Getter for the maxixmum number of channels allowed for the instance.
+   *
+   * @return {number} Buffer length in frames.
+   */
+  get maxChannelCount() {
+    return this._isInitialized ? this._maxChannelCount : null;
+  }
+
+  /**
+   * Returns a Float32Array object for a given channel index. If the channel
+   * index is undefined, it returns the reference to the base array of channel
+   * data.
+   *
+   * @param  {number|undefined} channelIndex Channel index.
+   * @return {Float32Array|Array[Float32Array]} a channel data array or an
+   * array of channel data.
+   */
+  getChannelData(channelIndex) {
+    console.assert(this._isInitialized);
+    console.assert(typeof channelIndex === 'undefined' ||
+                   channelIndex < this._channelCount);
+
+    return typeof channelIndex === 'undefined'
+        ? this._channelData : this._channelData[channelIndex];
+  }
+
+  /**
+   * Returns the base address of the allocated memory space in the WASM heap.
+   *
+   * @return {number} WASM Heap address.
+   */
+  getHeap() {
+    console.assert(this._isInitialized);
+
+    return this._dataPtr;
+  }
+
+  /**
+   * Frees the allocated memory space in the WASM heap.
+   *
+   * @return {[type]} [description]
+   */
+  free() {
+    console.assert(this._isInitialized);
+
+    this._isInitialized = false;
+    Module._free(this._dataPtr);
+    Module._free(this._pointerArrayPtr);
+    this._channelData = null;
+  }
+
+}  // class HeapAudioBuffer
+
+
+/**
```

Comment:
As a kind of general fifo utility, would it make sense to separate this out into its own file?

Reply:
This file is not supposed to be looked at by the article audience. I optimized the structure for the simplicity and did not want to dealt with the dependency between files as a part of the article content.

Eventually this can grow to a small library, but until then I would like to keep it simple.